### PR TITLE
Implement ArrayLiteral pop and shift

### LINKED ISSUE #6023: Feature request: ArrayLiteral#pop & #shift
`ArrayLiteral#push` works (although not documented), so I suppose it's possible to add `#pop` and `#shift` etc.

*Upd*
Nice place to start: https://github.com/crystal-lang/crystal/pull/5265

### PR DESCRIPTION
**Pull request by Blacksmoke16**
Resolve #6023

### CODE REVIEW

**Comment by vladfaust** [comment]
Why accepting NumberLiteral arguments only? 😕

File: src/compiler/crystal/macros.cr [hunk 1]
```diff
@@ -646,10 +646,18 @@ module Crystal::Macros
     def unshift : ArrayLiteral
     end
 
+    # Similar to `Array#shift`
+    def shift(n : NumberLiteral) : ArrayLiteral
```
Comment (Sija) [inline src/compiler/crystal/macros.cr]:
```suggestion
    def shift(n : ASTNode) : ArrayLiteral
```

File: src/compiler/crystal/macros.cr [hunk 2]
```diff
@@ -646,10 +646,18 @@ module Crystal::Macros
...
+    end
+
     # Similar to `Array#push`.
     def push(value : ASTNode) : ArrayLiteral
     end
 
+    # Similar to `Array#pop`
+    def pop(n : NumberLiteral) : ArrayLiteral
```
Comment (Sija) [inline src/compiler/crystal/macros.cr]:
```suggestion
    def pop(n : ASTNode) : ArrayLiteral
```

**Comment by Sija** [comment]
@vladfaust good catch, it should be `ASTNode` all the way.

**Comment by asterite** [comment]
Why `ASTNode`? 🤔

It's fine as `NumberLiteral`.

**Comment by Blacksmoke16** [comment]
@Sija Why tho?  `Array#shift` and pop are limited to integers.  `[1,2,3].pop("foo")` wouldn't make sense.

**Comment by vladfaust** [comment]
Yeah, my bad. This PR is okay

**Comment by Sija** [comment]
@Blacksmoke16 Sorry, I misread (was thinking about `#unshift`)

File: src/compiler/crystal/macros.cr [hunk 3]
```diff
@@ -646,10 +646,18 @@ module Crystal::Macros
     def unshift : ArrayLiteral
```
Comment (Sija) [inline src/compiler/crystal/macros.cr]: Shouldn't `#unshift` accept an argument?

Comment (Blacksmoke16) [reply src/compiler/crystal/macros.cr]: It does but just isn't documented it looks like.

Comment (Sija) [reply src/compiler/crystal/macros.cr]: So it should - question is whether in this PR or the new one?

Comment (Blacksmoke16) [reply src/compiler/crystal/macros.cr]: fixed in 201c834

Comment (vladfaust) [reply src/compiler/crystal/macros.cr]: I'd put into another PR, because one is about adding new features, and another is a fix

**Comment by bcardiff** [comment]
I think we should pursuit to remove the mutation methods over literal values. I stated that before.

The very nature of the mutation force a dependency of the execution order. If we aim for modular/incremental compilation that dependency would need to be tracked.

One of the first mutations `ArrayLiteral#push` introduced were to support registry-like feature. Pushing classes to an array for later iterate in a macro. This pattern can be substituted with other constructs (that don't exist yet) using annotations: "enumerate classes that has certain annotation applied".

I have see the usage of mutating literal like that. Append this for later iteration.

I would avoid adding more mutation if the reason is because other mutation operations exists.

Another benefit of the patter that use annotations is that it _would_ even be sound to semantically treat them as a declaration instead of an statement. That is, if a classes is later created with that annotation, the macro could be reevaluated; hence, there is no dependency of the order.

**Comment by Blacksmoke16** [comment]
@bcardiff :+1: works for me.  I'll make another PR for the current missing arg on `unshift` then.

**Comment by bcardiff** [comment]
I'm confused @Blacksmoke16 . Probably you have a use case that you need to unshift or pop the ArrayLiteral. Would you mind sharing it?

I get that adding an argument might be smaller change than adding an argument, but at the end of the day you seem to be needing something that I am not vouching in supporting. Again, can you share the use case?

**Comment by vladfaust** [comment]
My use-case was an ORM where I'd need to store meta data about models in a compile-time Array. Since I'm moving to annotations as well, I don't need this functionality either.

**Pull request closed without merging**